numpy/core/src/multiarray/getset.c

```diff
@@ -67,28 +67,32 @@ array_shape_set(PyArrayObject *self, PyObject *val)
         return -1;
     }
 
-    /* Free old dimensions and strides */
-    npy_free_cache_dim_array(self);
     nd = PyArray_NDIM(ret);
-    ((PyArrayObject_fields *)self)->nd = nd;
     if (nd > 0) {
         /* create new dimensions and strides */
-        ((PyArrayObject_fields *)self)->dimensions = npy_alloc_cache_dim(2 * nd);
-        if (PyArray_DIMS(self) == NULL) {
+        npy_intp *_dimensions = npy_alloc_cache_dim(2 * nd);
+        if (_dimensions == NULL) {
             Py_DECREF(ret);
             PyErr_SetString(PyExc_MemoryError,"");
             return -1;
         }
-        ((PyArrayObject_fields *)self)->strides = PyArray_DIMS(self) + nd;
+        /* Free old dimensions and strides */
+        npy_free_cache_dim_array(self);
+        ((PyArrayObject_fields *)self)->nd = nd;
+        ((PyArrayObject_fields *)self)->dimensions = _dimensions; 
+        ((PyArrayObject_fields *)self)->strides = _dimensions + nd;
+
         if (nd) {
             memcpy(PyArray_DIMS(self), PyArray_DIMS(ret), nd*sizeof(npy_intp));
             memcpy(PyArray_STRIDES(self), PyArray_STRIDES(ret), nd*sizeof(npy_intp));
         }
     }
     else {
+        ((PyArrayObject_fields *)self)->nd = 0;
```

Comment:
Doesn't this path need to free the old strides too?

Reply:
Yes, you're right. `npy_free_cache_dim_array(self);` should be called out side of the if-else block to cover both branches.

Reply:
@eric-wieser : Since I'm quite new to GitHub, I'm not sure what to do next? Should I change my commit or can you edit it for me?

Reply:
You can amend the previous commit and push back to your master branch.

Note that making PRs from master is usually a bad idea, as it's easy to forget you did it and then push something else on top.

Something to remember for next time.

Reply:
@eric-wieser : Thank you for the advice. I'll do it next time.